fix: Function.name failing IE tests

lucasfcosta opened the pull request:
Here is the fix for #3.

The reason why the build was failing on IE was because of `Function.name` which wasn't supported.
Now instead of getting a function's name using it's `name` I call a function called `getFunctionName` which runs a polyfill (which I've found [here](https://github.com/JamesMGreene/Function.name) and made some minor adaptations) if `.name` is not defined for that function, otherwise we just get it's name through the `.name` property.

This should be enough to work on every single browser.

I also noticed that SauceLabs tests aren't running for PRs. I'll take a look at it further this week too.

PS.: I couldn't add any tests for this since deleting the name property of a function or assigning `undefined` to it makes it `''` instead of `undefined` in order to run the polyfill. However, coverage will be 100% if we consider this branch runs on IE only. However, I'm open to suggestions if you can think of anything else for this.

meeber commented on index.js:
```diff
@@ -86,6 +86,36 @@ function compatibleMessage(thrown, errMatcher) {
+  var name = '';
+  if (typeof constructorFn.name === 'undefined') {
+    // Here we run a polyfill if constructorFn.name is not defined
+    var fnNamePrefixRegex = /^[\S\s]*?constructorFntion\s*/;
```
`constructorFntion`.... hmmm.... Did you do a search-and-replace of "func" to "constructorFn" at some point? :D :D :D 

(It does need to be changed back to `function` to work).

keithamus replied:
These regexes are a little redundant; `[\S\s]` will match any non-whitespace character, plus any whitespace character - so in other words... any character. In addition the `?` following the `*` is redundant. I also think using `replace` to strip out parts is a little cumbersome, when we can use `match` to extract the right bits. So something like:

```js
String(function foo(){}).match(/\s*function\s*([^\(]+)/)[1]
```

However I did a little testing, and it seems that IE9, at least, preserves comments in functions, so we should also look for those and strip them out where necessary. So something a bit more robust would be:

```js
String(function/*one*/ foo/*two*/(){}).match(/\s*function(?:\s*|\s*\/\*[^(?:*/)]+\*\/\s*)*?([^\(\/]+)/)[1]
```

Also (small aside) think it'd be worth caching the regexp as a var outside of the function - so it isn't instantiated each time. So ultimately this whole block becomes:

```js
var functionNameMatch = /\s*function(?:\s*|\s*\/\*[^(?:*/)]+\*\/\s*)*?([^\(\/]+)/;

function getFunctionName(constructor) {
  if (typeof constructor.name === 'undefined') {
    var match = String(constructor).match(functionNameMatch);
    if (match) {
      return match[1];
    }
    return '';
  }
  return constructor.name;
}
```

Of course it's probably wise to also add some unit tests which demonstrate that it successfully extracts the right name even with comments inline.

lucasfcosta replied:
@keithamus @meeber great catches, guys.
Did this late yesterday, so these things have slipped out.
It's always good to have you eyes on this.

(And nice regexp "magic" there Keith :smile: )

meeber commented on index.js:
```diff
@@ -99,10 +129,10 @@ function compatibleMessage(thrown, errMatcher) {
   } else if (typeof errorLike === 'function') {
     // if `err` is not an instance of Error it is an error constructor itself
-    constructorName = new errorLike().name; // eslint-disable-line new-cap
+    constructorName = getFunctionName(errorLike); // eslint-disable-line new-cap
```
Line 135: What was the purpose of the `new` in the previous version? Just want to verify it wasn't important, since that aspect is lost in the new version. Also, with the `new` gone, `// eslint-disable-line new-cap` is no longer needed.

lucasfcosta replied:
You seem to be right.
Apparently we didn't need `new` at all. I've just tested this with and without `new` and results are the same.

meeber commented:
@lucasfcosta Woooo more IE compatibility fun :D

I made a couple notes above.

I'm not too familiar with the sauce stuff yet, but @keithamus offered the following explanation when I asked about the lack of sauce tests on a Chai PR:

> Annoyingly, the saucelabs access keys have to be encrypted and Travis doesn't use encrypted variables for pull requests - meaning SauceLabs cannot run. Sadly until saucelabs provides first class integration with travis, similar to what coveralls has done, then we won't get saucelabs builds on PRs.

keithamus commented:
> I also noticed that SauceLabs tests aren't running for PRs. I'll take a look at it further this week too.

No point btw. Saucelabs doesn't integrate with Travis, and Travis doesn't allow encrypted values (for example a saucelabs access token) in PR builds, in case they are leaked. Until one of those facts changes, we're SOL sadly

keithamus commented:
> PS.: I couldn't add any tests for this since deleting the name property of a function or assigning undefined to it makes it '' instead of undefined in order to run the polyfill. However, coverage will be 100% if we consider this branch runs on IE only. However, I'm open to suggestions if you can think of anything else for this.

As part of the master build, the coverage reports for all browsers are combined into one. So this should come back as 100% coverage 😄

lucasfcosta commented:
Hey guys, thanks for spotting these.
Sorry for letting these things escape my mind.
I promise I'm gonna review things with increased attention next time. :smile:

keithamus commented:
@lucasfcosta Don't beat yourself up about it, the whole reason we do PRs is so that we can catch mistakes or make suggestions!

lucasfcosta commented:
I've just improved the RegExp as you guys indicated and added some tests for it, the implementation looks much more clear now. You can see them [here](https://github.com/chaijs/check-error/pull/4/files#diff-910eb6f57886ca16c136101fb1699231R88) and [here](https://github.com/chaijs/check-error/pull/4/files#diff-910eb6f57886ca16c136101fb1699231R89).

I also removed the `eslint-disable-line no-new-cap` comment since we're not using `new` anymore.

Thanks for your inputs, you guys were awesome (as you always do) :heart: 

Please let me know if I'm missing anything else.

keithamus commented:
LGTM. @meeber?

meeber commented on index.js:
```diff
@@ -86,6 +86,35 @@ function compatibleMessage(thrown, errMatcher) {
+    if (match) {
+      name = match[1];
+    }
+    name = '';
```
Line 109 overwrites the value that's set on line 107.

keithamus replied:
Oh yeah, good catch!

meeber commented:
Hey guys, I'm not able to test the actual code in IE at the moment, but just tried this small regex test in isolation:

```js
var functionNameMatch = /\s*function(?:\s*|\s*\/\*[^(?:*/)]+\*\/\s*)*?([^\(\/]+)/;

console.log("function withoutComments()".match(functionNameMatch));
console.log("function /*one*/correctName/*two*/()".match(functionNameMatch));
```

The first one matched "withoutComments", but the second one didn't match anything.

keithamus commented:
RegExp should be `\s*function(?:\s|\s*\/\*[^(?:*\/)]+\*\/\s*)*([^\(\/]+)`

lucasfcosta commented:
Here it is.
Well, I've tested @meeber's examples and I've also checked this regexp's rules using [this tool](https://regex101.com/#javascript). 

I also fixed that incorrect assignment to `name` on L109.

meeber commented:
@lucasfcosta Nicely done, sir. LGTM!
